File: POF/Erros.fs
module Erros

type ErroInsercaoPaleta =
    | ErroAoLerArquivo of nomeDoArquivo: string * mensagemDeErro: string
    | FormatoDePaletaInvalido of linhas: (int * string) array
    | ValorDeCorInvalido of linhas: (int * string) array
    | ErroAoSalvarPaletaNoBancoDeDados of mensagem: string

    member this.MensagemDeErro() =
        match this with
        | ErroAoLerArquivo(nomeDoArquivo, mensagemDeErro) -> $"Erro ao ler arquivo '{nomeDoArquivo}': {mensagemDeErro}"
        | FormatoDePaletaInvalido(linhas) ->
            let linhasErradas =
                linhas
                |> Array.map (fun (numero, texto) -> $"\tLinha {numero + 1}: {texto}")
                |> String.concat "\n"

            $"Formato de paleta inválido:\n{linhasErradas}"
        | ValorDeCorInvalido(linhas) ->
            let linhasErradas =
                linhas
                |> Array.map (fun (numero, texto) -> $"\tLinha {numero + 1}: {texto}")
                |> String.concat "\n"

            $"Valor de cor inválido (insira valores entre 0 e 255):\n{linhasErradas}"
        | ErroAoSalvarPaletaNoBancoDeDados(mensagem) ->
            $"Erro ao persistir paleta no banco de dados: {mensagem}\n\nSe persistir o erro, contate o administrador do sistema"

File: ts/src/erros.ts
export type ErroInsercaoPaleta =
  | { kind: 'erro_ao_ler_arquivo', nomeDoArquivo: string, mensagemDeErro: string }
  | { kind: 'formato_de_paleta_invalido', linhas: [number, string][] }
  | { kind: 'valor_de_cor_invalido', linhas: [number, string][] }
  | { kind: 'erro_ao_salvar_paleta_no_banco_de_dados', mensagemDeErro: string }

export const mensagemErroInsercaoPaleta = (erro: ErroInsercaoPaleta) => {
  switch (erro.kind) {
    case 'erro_ao_ler_arquivo':
      return `Erro ao ler arquivo ${erro.nomeDoArquivo}: ${erro.mensagemDeErro}`

    case 'formato_de_paleta_invalido':
      const linhasPaletaInvalidas = erro.linhas.map(([linha, valor]) => `\t${linha + 1}: ${valor}`).join('\n')
      return `Formato de paleta inválido:\n${linhasPaletaInvalidas}`

    case 'valor_de_cor_invalido':
      const coresInvalidas = erro.linhas.map(([linha, valor]) => `\t${linha + 1}: ${valor}`).join('\n')
      return `Valor de cor inválido, insira valores entre 0 e 255:\n${coresInvalidas}`

    case 'erro_ao_salvar_paleta_no_banco_de_dados':
      return `Erro ao salvar paleta no banco de dados: ${erro.mensagemDeErro}\n
      Se persistir o erro, contate o administrador do sistema.`
  }
}